Exceptions

JeremieLariviere opened the pull request.

JeremieLariviere committed:
refactor: Move ExceptionHandling to infra

JeremieLariviere committed:
refactor: Remove unused temporary file

JeremieLariviere committed:
refactor: Remove hardcoded statuscodes

JeremieLariviere committed:
Merge branch 'dev' into exceptions

JeremieLariviere committed:
refactor: Extract error controller extension method to infra

nicolascnbeslogic commented on Infra.ExceptionHandling/Infra.ExceptionHandling.Controllers/GenericWebExceptionHandler.cs:
```diff
@@ -0,0 +1,56 @@
+    public GenericWebExceptionHandler(IServiceProvider serviceProvider, IWebHostEnvironment env)
+    {
+      this.serviceProvider = serviceProvider;
+      showDetails = env.IsDevelopment();
```
Extract env.IsDevelopment ?

nicolascnbeslogic commented on Infra.ExceptionHandling/Infra.ExceptionHandling.Extensions/GenericTypeUtils.cs:
```diff
@@ -0,0 +1,22 @@
+{
+  internal static class GenericTypeUtils
+  {
+    internal static IEnumerable<Type> BaseTypes(this Type type)
```
Peut-être le nommer "AncestorsBaseTypes" ? Ça m'a pris du temps à lire et comprendre.
Peut-être même une autre implémentation plus verbose ?

```
public static IEnumerable<Type> BaseTypesRecursive(this Type type)
    {
      var baseType = type?.BaseType;
      if (baseType != null)
      {
        yield return baseType;
        foreach (var parentBaseType in BaseTypesRecursive(baseType))
        {
          yield return parentBaseType;
        }
      }
    }
```

nicolascnbeslogic reviewed (commented):
C'est all good ! Je t'ai proposé des petites affaires, mais tu prends ce que tu veux et tu laisse ce que tu veux.

Prêt à merger

pedro-beslogic commented on Infra.ExceptionHandling/Infra.ExceptionHandling.Controllers/ErrorController.cs:
```diff
@@ -0,0 +1,27 @@
+    // This feature gets added to the HttpContext by Microsoft's error handling 
+    // middleware before calling the error route so it should always be present
+    // hence the ! null-forgiving operator
+    private Exception GrabError() => HttpContext.Features.Get<IExceptionHandlerPathFeature>()!.Error;
```
Not sure why but I feel that `GrabError` could be a better naming such as `HandleError`, `ProcessError`, something like that. What do you guys think?

JeremieLariviere replied:
my reasoning was that it quite litteraly grabs the error from the HttpContext. It doesn't do any handling

JeremieLariviere committed:
refactor: Rename GenericWebExceptionHandler to GlobalExceptionHandler

JeremieLariviere committed:
refactor: Extract webHost from exception handler

it was only used to check for development environment. so i pass the boolean directly

JeremieLariviere committed:
feat: Add a default domain exception handler

this is so we don't have to implement a handler for every error.
